nixos/modules/programs/command-not-found/src/command-not-found.cpp

```diff
@@ -0,0 +1,141 @@
+    if (db) {
+      sqlite3_close(db);
+    }
+  }
+};
+
+struct SqliteStmtDeleter {
+  void operator()(sqlite3_stmt *const stmt) {
+    if (stmt) {
+      sqlite3_finalize(stmt);
+    }
+  }
+};
+
+int queryPackages(const char *system, const char *program,
+                  std::vector<std::string> &packages) {
+  sqlite3 *raw_db = nullptr;
+  if (sqlite3_open(DB_PATH, &raw_db) != SQLITE_OK) {
+    fprintf(stderr, "cannot open database \"%s\": %s\n", DB_PATH,
+            sqlite3_errmsg(raw_db));
+  };
+  std::unique_ptr<sqlite3, SqliteDeleter> db(raw_db);
+
+  const char *sql =
+      "select package from Programs where system = ? and name = ?;";
+
+  sqlite3_stmt *raw_stmt = nullptr;
+  if (sqlite3_prepare_v2(db.get(), sql, -1, &raw_stmt, NULL) != SQLITE_OK) {
+    fprintf(stderr, "Failed to prepare query: %s\n", sqlite3_errmsg(db.get()));
+    return 1;
+  }
+  std::unique_ptr<sqlite3_stmt, SqliteStmtDeleter> stmt(raw_stmt);
+
+  if (sqlite3_bind_text(stmt.get(), 1, system, -1, 0) != SQLITE_OK) {
+    fprintf(stderr, "Failed to bind system parameter: %s\n",
+            sqlite3_errmsg(db.get()));
+    return 1;
+  };
+  if (sqlite3_bind_text(stmt.get(), 2, program, -1, 0) != SQLITE_OK) {
+    fprintf(stderr, "Failed to bind program parameter: %s\n",
+            sqlite3_errmsg(db.get()));
+    return 1;
+  };
+
+  while (sqlite3_step(stmt.get()) == SQLITE_ROW) {
+    std::string pkgs = (const char *)(sqlite3_column_text(stmt.get(), 0));
+    packages.push_back(std::move(pkgs));
+  }
+
+  return 0;
+}
+
+int main(int argc, char **argv) {
+  const char *system = getenv("NIX_SYSTEM");
+  const char *program = argv[1];
+
+  if (!system) {
+    system = NIX_SYSTEM;
+  }
+
+  if (argc < 2) {
+    fprintf(stderr, "USAGE: %s PROGRAM\n", argv[0]);
+    return 1;
+  }
+
+  std::vector<std::string> packages = {};
+  if (queryPackages(system, program, packages) < 0) {
+    return 1;
+  };
+
+  switch (packages.size()) {
+  case 0: {
+    fprintf(stderr, "%s: command not found\n", program);
+    break;
+  }
+  case 1: {
+    const std::string package = packages[0];
+    if (getenv("NIX_AUTO_INSTALL")) {
+      fprintf(stderr,
+              "The program '%s' is currently not installed. It is provided by"
+              "the package '%s', which I will now install for you.\n",
+              program, package.c_str());
+      pid_t pid = 0;
+      std::string attrname = std::string("nixos.") + program;
+      const char *args[4] = {"nix-env", "-iA", &attrname[0], nullptr};
+      int status = posix_spawnp(&pid, "nix-env", nullptr, nullptr,
+                                const_cast<char **const>(args), environ);
+      if (status == 0 && waitpid(pid, &status, 0) != -1) {
+        if (WIFEXITED(status) && WEXITSTATUS(status) == 0) {
+          return 126;
+        }
+      }
+    } else if (getenv("NIX_AUTO_RUN")) {
+      std::string attrname = std::string("nixpkgs.") + package;
+      std::vector<const char *> args = {"nix", "run", &attrname[0],
```

Comment:
`nix run` should be avoided, it's experimental and won't work in Nix 2.4 unless you pass `--experimental-features nix-command`.

Reply:
@edolstra is this going away in future?

Reply:
FYI, `nix run` is used a few places in nixpkgs (and the manual) already.

Reply:
Not only `nix run` also `nix copy` and `nix build` (for example in the installer).

Reply:
Yeah, we should replace those by `nix-build` etc. (or add `--experimental-features nix-command` to those invocations, but that's tricky because it could override the user's experimental-features setting).

Reply:
BTW, maybe we should get rid of `NIX_AUTO_RUN` as well, automatically downloading/running binaries is not a very secure thing to do...

Reply:
Let's say not safe instead of secure.